src/mgr/ActivePyModules.cc

```diff
@@ -1516,16 +1516,42 @@ void ActivePyModules::register_client(std::string_view name, std::string addrs)
   entity_addrvec_t addrv;
   addrv.parse(addrs.data());
 
-  dout(7) << "registering msgr client handle " << addrv << dendl;
-  py_module_registry.register_client(name, std::move(addrv));
+  dout(7) << "registering msgr client handle [" << name << "] " << addrv << dendl;
+  epoch_t e;
+  if (cluster_state.is_client_in_mgrmap(name, addrv, &e)) {
+    // Client already registered in the MgrMap
+      return;
+  }
+  py_module_registry.register_client(name, addrv);
+  while (true) {
+    // Wait to receive MgrMap from the monitor
+    C_SaferCond c;
+    cluster_state.wait_for_mgrmap(&c, e+1);
+    c.wait();
```

Comment:
@idryomov I've locally made the other straight forward changes requested. Just want to make sure if you're suggesting the following,
```
diff --git a/src/mgr/ActivePyModules.cc b/src/mgr/ActivePyModules.cc
index 3d61065cd80..5de98afec70 100644
--- a/src/mgr/ActivePyModules.cc
+++ b/src/mgr/ActivePyModules.cc
@@ -1518,31 +1518,28 @@ void ActivePyModules::register_client(std::string_view name, std::string addrs)
 
   dout(7) << "registering msgr client handle [" << name << "] " << addrv << dendl;
   epoch_t e;
-  if (cluster_state.is_client_in_mgrmap(name, addrv, &e)) {
-    // Client already registered in the MgrMap
-      return;
-  }
   py_module_registry.register_client(name, addrv);
   while (true) {
-    // Wait to receive MgrMap from the monitor
-    C_SaferCond c;
-    cluster_state.wait_for_mgrmap(&c, e+1);
-    c.wait();
     if (cluster_state.is_client_in_mgrmap(name, addrv, &e)) {
-      // Client registered in the updated MgrMap
+      // Client registered in the MgrMap
       return;
     }
     // While waiting to receive new MgrMap from the monitor, the
-    // client may be deregistered from the py_module_registry. If this
-    // occurs, the client may never be registered in the MgrMap and the
-    // caller would be stuck trying to register a client.
+    // client may be deregistered from the PyModuleRegistry (perhaps by another
+    // module replacing e.g. libcephsqlite address). If this  occurs, the
+    // client may never be registered in the MgrMap and the caller would be
+    // stuck trying to register a client.
     if (!py_module_registry.is_client_in_registry(name, addrv)) {
       derr << "msgr client handle [" << name << "] " << addrv
-           << " deregistered from py_module_registry "
+           << " deregistered from PyModuleRegistry "
            << "when registering in MgrMap"
            << dendl;
       throw std::runtime_error("Failed to register client");
     }
+    // Wait to receive MgrMap from the monitor
+    C_SaferCond c;
+    cluster_state.wait_for_mgrmap(&c, e+1);
+    c.wait();
   }
 }
```

Reply:
I think the `cluster_state.is_client_in_mgrmap` check at the top (outside of the loop) should remain.  IIUC Patrick wanted it in place to prevent adding a duplicate to the registry: https://github.com/ceph/ceph/pull/51169#discussion_r1186131462

Otherwise, yes, something like that.

Reply:
With the restructured loop, you still have this problem:

>  For example, if the client gets deregistered from PyModuleRegistry immediately after it's registered, we would be waiting for the next mgrmap for nothing here.

The unregister call may still occur after the `is_client_in_registry` call. If no beacon is sent between `py_module_registry.register_client(name, addrv);` and `py_module_registry.is_client_in_registry` then you still wait for the next mgrmap "for nothing".

The only way to fix this, I think, is to prevent unregistering a client until the mgrmap includes the client addr. But, this is a reallllllly obscure corner case where the only cost is a thread unnecessarily blocking for some time on the next mgrmap.

Reply:
> With the restructured loop, you still have this problem:

Yeah, without locking (not really possible here) or something along the lines of what you described to compensate, there is always going to be race.  It's just a personal experience that loops that sleep/wait first are more error prone.  
Ramana, feel free to disregard the restructuring suggestion and just amend the comment.